packages/block-editor/src/store/reducer.js

```diff
@@ -1237,54 +1237,105 @@ function selection( state = {}, action ) {
 }
 
 /**
- * Reducer returning the block selection's start.
+ * Higher-order reducer that ensures a valid block selection remains after the
+ * RESET_BLOCKS action.
  *
- * @param {Object} state  Current state.
- * @param {Object} action Dispatched action.
+ * @param {Function} reducer The selection reducer.
  *
- * @return {Object} Updated state.
+ * @return {Object} Selection state.
  */
-export function selectionStart( state = {}, action ) {
-	switch ( action.type ) {
-		case 'SELECTION_CHANGE':
+const withSelectionUpdateOnBlockReset = ( reducer ) => ( state, action ) => {
+	// When RESET_BLOCKS is dispatched, the currently selected blocks may be
+	//  replaced, in which case selection should be cleared.
+	if ( action.type === 'RESET_BLOCKS' ) {
+		const startClientId = state?.selectionStart?.clientId;
+		const endClientId = state?.selectionEnd?.clientId;
+
+		// Do nothing if there's no selected block.
+		if ( ! startClientId && ! endClientId ) {
+			return state;
+		}
+
+		// If the start of the selection won't exist after reset, remove selection.
+		if (
+			! action.blocks.some(
+				( block ) => block.clientId === startClientId
+			)
+		) {
 			return {
-				clientId: action.clientId,
-				attributeKey: action.attributeKey,
-				offset: action.startOffset,
+				selectionStart: {},
+				selectionEnd: {},
 			};
-		case 'RESET_SELECTION':
-			return action.selectionStart;
-		case 'MULTI_SELECT':
-			return { clientId: action.start };
-	}
-
-	return selection( state, action );
-}
+		}
 
-/**
- * Reducer returning the block selection's end.
- *
- * @param {Object} state  Current state.
- * @param {Object} action Dispatched action.
- *
- * @return {Object} Updated state.
- */
-export function selectionEnd( state = {}, action ) {
-	switch ( action.type ) {
-		case 'SELECTION_CHANGE':
+		// If the end of the selection won't exist after reset, collapse selection.
+		if (
+			! action.blocks.some( ( block ) => block.clientId === endClientId )
+		) {
 			return {
-				clientId: action.clientId,
-				attributeKey: action.attributeKey,
-				offset: action.endOffset,
+				...state,
+				selectionEnd: state.selectionStart,
 			};
-		case 'RESET_SELECTION':
-			return action.selectionEnd;
-		case 'MULTI_SELECT':
-			return { clientId: action.end };
+		}
 	}
 
-	return selection( state, action );
-}
+	return reducer( state, action );
+};
+
+export const selection = flow(
+	combineReducers,
+	withSelectionUpdateOnBlockReset
+)( {
+	/**
+	 * Reducer returning the block selection's start.
+	 *
+	 * @param {Object} state  Current state.
+	 * @param {Object} action Dispatched action.
+	 *
+	 * @return {Object} Updated state.
+	 */
+	selectionStart( state = {}, action ) {
```

Comment:
Do you think it should be just "start" and "end"  to avoid the redundancy

Reply:
@youknowriad I've tried that out in ead6d1a. It also allowed for refactoring the higher-order reducer into the main selection reducer.

I don't think it resulted in much reduced repetition, but I like that the code is a little easier to follow now that there's no higher order reducer.